common/lib/capa/capa/responsetypes.py

```diff
@@ -831,26 +854,146 @@ def assign_choice_names(self):
+        If one error, half credit as long as there are 3+ choices
+        If two errors, 1/4 credit as long as there are 5+ choices
+        (If not enough choices, no credit.)
+        Arguments: all_choices, student_answer, student_non_answers
+        Returns a CorrectMap
+        """
+
+        all_choices = kwargs['all_choices']
+        student_answer = kwargs['student_answer']
+        student_non_answers = kwargs['student_non_answers']
+
+        halves_error_count = 0
+
+        incorrect_answers = sum([1 for answer in student_answer if answer in self.incorrect_choices])
+        missed_answers = sum([1 for blank in student_non_answers if blank in self.correct_choices])
+        halves_error_count = incorrect_answers + missed_answers
+
+        if halves_error_count == 0:
+            return_grade = self.get_max_score()
+            return CorrectMap(self.answer_id, correctness='correct', npoints=return_grade)
+        elif halves_error_count == 1 and len(all_choices) > 2:
+            return_grade = round(self.get_max_score() / 2.0, 2)
+            return CorrectMap(self.answer_id, correctness='partially-correct', npoints=return_grade)
+        elif halves_error_count == 2 and len(all_choices) > 4:
+            return_grade = round(self.get_max_score() / 4.0, 2)
+            return CorrectMap(self.answer_id, correctness='partially-correct', npoints=return_grade)
+        else:
+            return CorrectMap(self.answer_id, 'incorrect')
+
+    def grade_without_partial_credit(self, **kwargs):
+        """
+        Standard grading for checkbox problems.
+        100% credit if all choices are correct; 0% otherwise
+        Arguments: student_answer
+        """
+
+        student_answer = kwargs['student_answer']
 
         required_selected = len(self.correct_choices - student_answer) == 0
         no_extra_selected = len(student_answer - self.correct_choices) == 0
 
-        correct = required_selected & no_extra_selected & no_empty_answer
+        correct = required_selected & no_extra_selected
 
         if correct:
             return CorrectMap(self.answer_id, 'correct')
         else:
             return CorrectMap(self.answer_id, 'incorrect')
 
+    def get_score(self, student_answers):
+
+        # Setting up answer lists:
+        #  all_choices: the full list of checkboxes
+        #  student_answer: what the student actually chose (note no "s")
+        #  student_non_answer: what they didn't choose
+        #  self.correct_choices: boxes that should be checked
+        #  self.incorrect_choices: boxes that should NOT be checked
+
+        all_choices = self.correct_choices.union(self.incorrect_choices)
+
+        student_answer = student_answers.get(self.answer_id, [])
+
+        if not isinstance(student_answer, list):
+            student_answer = [student_answer]
+
+        # "None apply" should really be a valid choice for "check all that apply",
+        # but score feedback is broken when no boxes are marked.
```

Comment:
This comment is a little confusing because it's not quite clear what the current behavior is. When a student doesn't mark any boxes, their answer will be marked as incorrect, right?